### PR TITLE
Merge Avalara tax into v5 0

### PR DESCRIPTION
**Pull request by ChristopherCotnoir**

### CODE REVIEW

File: foundation-database/public/functions/posttax.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    IF (SELECT getOrderTax('VCH', vohead_id) - vohead_tax_charged <= 0.0
+          FROM vohead
+         WHERE vohead_id = pOrderId) THEN
+      RETURN NULL;
+    END IF;
+  END IF;
+
+  _service := fetchMetricText('TaxService');
+
+  IF _service = 'A' THEN
```
Comment (andersdd) [inline foundation-database/public/functions/posttax.sql]: Consider `IF (fetchMetricText('TaxService') = 'A') THEN` to simplify code

File: foundation-database/public/functions/refundtax.sql [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  PERFORM createARCreditMemo(_aropenid, invchead_cust_id, _number, '', pRefundDate,
+                             getOrderTax('INV', pInvcheadId), '', -1, -1, -1, pRefundDate, -1,
+                             invchead_salesrep_id, 0, NULL, invchead_curr_id)
+     FROM invchead
+    WHERE invchead_id = pInvcheadid;
+  
+  _service := fetchMetricText('TaxService');
+
+  IF _service = 'A' THEN
```
Comment (andersdd) [inline foundation-database/public/functions/refundtax.sql]: Consider `IF (fetchMetricText('TaxService') = 'A') THEN` to simplify code.  You can remove the variable `_service`

File: foundation-database/public/functions/updatemiscvouchertax.sql [hunk 3]
```diff
@@ -1,84 +0,0 @@
...
-
-   END LOOP taxtypes;
-
-   -- Tidy up
-   DROP TABLE IF EXISTS _vodisttax;
-   
-   -- All done   
-   RETURN _total;
-    
-END;
```
Comment (andersdd) [inline foundation-database/public/functions/updatemiscvouchertax.sql]: I hope this function is being replaced by logic elsewhere for VAT/GST countries.

File: foundation-database/public/tables/expcat.sql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+SELECT xt.add_constraint('expcat', 'expcat_expcat_code_key', 'UNIQUE (expcat_code)', 'public'),
+       xt.add_constraint('expcat', 'expcat_expcat_code_check', 'CHECK ((expcat_code <> ''::text))', 'public');
+
+ALTER TABLE public.expcat ENABLE TRIGGER ALL;
+
+COMMENT ON TABLE expcat IS 'Expense Category information';
+
+UPDATE expcat
+   SET expcat_tax_accnt_id = expcat_exp_accnt_id;
```
Comment (andersdd) [inline foundation-database/public/tables/expcat.sql]: This does not look idempotent.  What happens during the next release?

**Comment by andersdd** [comment]
CI failing due to `ERROR:  function selectforbilling(integer, numeric, boolean, integer) does not exist`

File: database/source/xt/functions/co_tax_total.sql [hunk 5]
```diff
@@ -1,10 +1,8 @@
...
-    group by taxdetail_tax_id) as data;
-$$ language sql;
+DROP function IF EXISTS xt.co_tax_total(cohead) CASCADE;
+  
+create or replace function xt.co_tax_total(cohead) 
+returns numeric as $$
+BEGIN
+  RETURN 0.00; -- DEPRECATED taxation function;
+END;
+$$ language plpgsql;
```
[the remark this answers is not in the document]
Comment (ChristopherCotnoir) [reply database/source/xt/functions/co_tax_total.sql]: I don't know much about the web client stuff so I'd leave it be as we're in the process of removing it all anyway.

File: foundation-database/public/functions/posttax.sql (hunk 1 above)
Comment (ChristopherCotnoir) [reply foundation-database/public/functions/posttax.sql]: I used this form to make potentially adding other services in the future easier.

File: foundation-database/public/functions/updatemiscvouchertax.sql (hunk 3 above)
Comment (ChristopherCotnoir) [reply foundation-database/public/functions/updatemiscvouchertax.sql]: The voucher system was completely reworked. Misc vouchers have an actual tax field now, I believe that handles it somehow.

File: foundation-database/public/tables/expcat.sql (hunk 4 above)
Comment (ChristopherCotnoir) [reply foundation-database/public/tables/expcat.sql]: Good catch, should be fixed now.

**Comment by ChristopherCotnoir** [comment]
jenkins - retest this please

**Comment by ChristopherCotnoir** [comment]
Ugh. I have the migration script creating a temp table for dochead/docitem views for performance reasons and designated them as ON COMMIT DROP, but since npm run-script test-build runs the foundation-database upgrade twice in the same transaction, that doesn't work. I'll have to change it to manual DROPs.

**Comment by ChristopherCotnoir** [comment]
Added one last bugfix that I'd had in the back of my head for a while but hadn't had the chance to put in.

**Pull request merged**